File: core/src/main/java/com/geometryduel/game/Body.java
package com.geometryduel.game;

import com.badlogic.gdx.math.Vector2;

/** 还原 pama1234.app.game.server.duel.util.Body。 */
public abstract class Body {
    public final Vector2 pos = new Vector2();
    public final Vector2 vel = new Vector2();
    public float directionAngle;
    public float speed;

    public void update() {
        pos.add(vel);
    }

    public void vel(float angle, float speed) {
        this.directionAngle = angle;
        this.speed = speed;
        this.vel.set(speed * cos(angle), speed * sin(angle));
    }

    public float dist(Body o) {
        return Vector2.dst(pos.x, pos.y, o.pos.x, o.pos.y);
    }

    public float distPow2(Body o) {
        float dx = o.pos.x - pos.x, dy = o.pos.y - pos.y;
        return dx * dx + dy * dy;
    }

    public float angle(Body o) {
        return (float) Math.atan2(o.pos.y - pos.y, o.pos.x - pos.x);
    }

    public abstract void display(com.geometryduel.render.Shapes s);

    public static float cos(float a) {
        return (float) Math.cos(a);
    }

    public static float sin(float a) {
        return (float) Math.sin(a);
    }
}

File: app/src/main/java/com/geometryduel/game/Body.kt
package com.geometryduel.game

import com.geometryduel.render.GameRenderer
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/** 二维向量（替代 libGDX Vector2）。 */
class Vec2 {
    var x = 0f
    var y = 0f

    fun set(x: Float, y: Float): Vec2 {
        this.x = x
        this.y = y
        return this
    }

    fun set(o: Vec2): Vec2 = set(o.x, o.y)

    fun add(o: Vec2): Vec2 {
        x += o.x
        y += o.y
        return this
    }

    fun scl(s: Float): Vec2 {
        x *= s
        y *= s
        return this
    }
}

/** 还原 pama1234.app.game.server.duel.util.Body。 */
abstract class Body {
    val pos = Vec2()
    val vel = Vec2()
    var directionAngle = 0f
    var speed = 0f

    open fun update() {
        pos.add(vel)
    }

    fun vel(angle: Float, speed: Float) {
        this.directionAngle = angle
        this.speed = speed
        this.vel.set(speed * cos(angle), speed * sin(angle))
    }

    fun dist(o: Body): Float {
        val dx = o.pos.x - pos.x
        val dy = o.pos.y - pos.y
        return sqrt(dx * dx + dy * dy)
    }

    fun distPow2(o: Body): Float {
        val dx = o.pos.x - pos.x
        val dy = o.pos.y - pos.y
        return dx * dx + dy * dy
    }

    fun angle(o: Body): Float {
        return kotlin.math.atan2(o.pos.y - pos.y, o.pos.x - pos.x)
    }

    abstract fun display(s: GameRenderer)

    companion object {
        fun cos(a: Float) = kotlin.math.cos(a)
        fun sin(a: Float) = kotlin.math.sin(a)
    }
}
